ResearchKit/Common/ORKSurveyAnswerCellForText.m

```diff
@@ -327,10 +340,22 @@ - (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRang
 }
 
 - (BOOL)textFieldShouldEndEditing:(UITextField *)textField {
+    BOOL isValid = [self isAnswerValid];
+    if (! isValid) {
+        [self showValidityAlertWithMessage:[[self.step impliedAnswerFormat] localizedInvalidValueStringWithAnswerString:textField.text]];
+        return NO;
+    }
+    
     return YES;
 }
 
 - (BOOL)textFieldShouldReturn:(UITextField *)textField {
+    BOOL isValid = [self isAnswerValid];
+    if (! isValid) {
+        [self showValidityAlertWithMessage:[[self.step impliedAnswerFormat] localizedInvalidValueStringWithAnswerString:textField.text]];
+        return NO;
+    }
+    
     [self.textField resignFirstResponder];
     return YES;
```

Comment:
Consolidate this code into a method, since it is the same for `textFieldShouldEndEditing`.

Also, the validity alert is not showing if invalid input is provided in `ORKTest` or `ORKCatalog`.